### ISSUE TITLE
mmsd2性能好像复现差点，不知道哪里失误了

### ISSUE THREAD

**Issue by Issac304** [Aichiniuroumian/RCLMuFN#3]
mmsd2性能好像复现差点，不知道哪里失误了

**Comment by SongshuL3** [comment]
The model incorrectly treats the **Batch dimension** as the **Sequence dimension** during the Cross-Attention mechanism in the `RCLM` and `CLIP-View Fusion` modules. This results in the inference of a specific sample being dependent on other random samples within the same batch. This behavior violates the **Independent and Identically Distributed** assumption during inference and contradicts the physical meaning of "cross-modal interaction" described in the paper.

Consequently, the model's performance is heavily reliant on the **Batch Size** and the statistical distribution of the specific batch, making it unstable and conceptually invalid for standard binary classification tasks.

## Implementation

The issue lies in `model.py`, specifically within the `CrossAttention` class and its usage in `RCLMuFN`.

**Location:** `model.py`

```python
class CrossAttention(nn.Module):
    def forward(self, query, key, value):
        # ......
        
        # If Inputs query and key are 2D tensors: [Batch_Size, Hidden_Dim], the transpose operation turns key into [Hidden_Dim, Batch_Size]
        attention_scores = torch.matmul(query, key.transpose(-1, -2)) 
        # Resulting attention_scores shape: [Batch_Size, Batch_Size]
        
        attention_weights = F.softmax(attention_scores, dim=-1)
        # The softmax normalizes across the batch dimension.
        
        attended_values = torch.matmul(attention_weights, value)
        return attended_values
```

**Usage in `RCLMuFN.forward`:**
```python
# Lines 174-175
txt_out = self.cross_att(txt_output, image_output, image_output)   # 32,768
# Lines 180-181 (CLIP-View Fusion)
cross_feature_text = self.cross_att(text_feature, image_feature, image_feature)  # 32,768
```

Here, `txt_output` and `text_feature` follow these pipelines:

```
# Line 132
# 'text_features' ARE NEVER USED. Pooling feature 'text_feature' are used instead.
text_features = output['text_model_output']['last_hidden_state']  # 128，77，512
# Line 134
text_feature = output['text_model_output']['pooler_output'] # 64，512

# Line 153-154
# 'last_hidden_states' ARE NEVER USED. Pooling feature 'pooler_outputs' are used instead.
last_hidden_states = outputs_bert.last_hidden_state  # 64,56,768
pooler_outputs = outputs_bert.pooler_output  # 64,768

# Line 165-166
# `squeeze` explicitly discards the sequence dimension.
txt_output = self.text_cross(tgt=text_feature2, memory=txt_cat)  # 32,1,768
txt_output = txt_output.squeeze(1)
```

Because the inputs are pooled global vectors (`[B, D]`) rather than sequences (`[B, L, D]`), the attention mechanism calculates the similarity between Sample $i$'s text and Sample $j$'s image (where $j$ ranges over the entire batch).

*   **Conceptual Implication:** The feature representation for Sample $A$ is a weighted sum of the images of Sample $A, B, C, ...$ in the current batch.
*   **Inference Violation:** In a real-world scenario (or Batch Size = 1), a tweet's sarcasm should be determined solely by its own text and image. However, this model requires a "batch context" to function as trained.


This implementation directly contradicts the methodology described in the paper.

*   **Paper Claim (Section 4.3):** *"We use two cross-attention modules to compute the interactions **between the modalities**... allows for a more comprehensive consideration of multimodal information."*
*   **Reality:** The interaction is **Inter-sample** (Text of Sample $i$ $\leftrightarrow$ Images of Batch $0...N$). The paper does not mention using batch-level statistics or contrastive mechanisms that would justify this behavior.

## Experiment


If the model learns true features, Test BS should not affect accuracy. If it relies on batch interaction, performance will collapse when Test BS differs from Train BS (especially at BS=1).

Results (All Experiments are conducted under original experiment setting in `main.py` and `train.sh`):

| Training Setting | Testing Setting | Acc (%) | F1 (%) | Observation |
| :--- | :--- | :--- | :--- | :--- |
| BS = 32          | BS = 32         | 88.54        | 88.44        | Reported setting in the paper and `train.sh`. Lower than results reported in the paper, but still much higher than baseline models. |
| BS = 32          | BS = 2          | 85.22        | 85.16        | Significant degradation                                 |
| **BS = 32**      | **BS = 1**      | **80.24**    | **80.22**    | Significant degradation. |
| BS = 128         | BS = 128        | 82.36 | 82.34 | Significant degradation |
| BS = 128         | BS = 32         | 81.53 | 81.43 | Significant degradation                                 |

When Batch Size = 1, `softmax` over the batch dimension results in a weight of 1.0 for the single sample, completely removing the "interaction" capability the model learned to rely on during training.

## Conclusion

**Avoid Using this Code as a Baseline**. Since the model violates the basic I.I.D. (Independent and Identically Distributed) assumption for inference, building upon this architecture may lead to theoretically unsound research. Any performance metrics reported with a Batch Size > 1 should not be considered representative of the model's actual predictive power.  Attempting to reproduce or build upon these results is likely to yield unstable and misleading outcomes, as the model's reported superiority relies on a methodological flaw that cannot be generalized to standard application scenarios.

**Comment by Aichiniuroumian** [comment]
> The model incorrectly treats the **Batch dimension** as the **Sequence dimension** during the Cross-Attention mechanism in the `RCLM` and `CLIP-View Fusion` modules. This results in the inference of a specific sample being dependent on other random samples within the same batch. This behavior violates the **Independent and Identically Distributed** assumption during inference and contradicts the physical meaning of "cross-modal interaction" described in the paper.
> 
> Consequently, the model's performance is heavily reliant on the **Batch Size** and the statistical distribution of the specific batch, making it unstable and conceptually invalid for standard binary classification tasks.
> 
> ## Implementation
> The issue lies in `model.py`, specifically within the `CrossAttention` class and its usage in `RCLMuFN`.
> 
> **Location:** `model.py`
> 
> class CrossAttention(nn.Module):
>     def forward(self, query, key, value):
>         # ......
>         
>         # If Inputs query and key are 2D tensors: [Batch_Size, Hidden_Dim], the transpose operation turns key into [Hidden_Dim, Batch_Size]
>         attention_scores = torch.matmul(query, key.transpose(-1, -2)) 
>         # Resulting attention_scores shape: [Batch_Size, Batch_Size]
>         
>         attention_weights = F.softmax(attention_scores, dim=-1)
>         # The softmax normalizes across the batch dimension.
>         
>         attended_values = torch.matmul(attention_weights, value)
>         return attended_values
> **Usage in `RCLMuFN.forward`:**
> 
> # Lines 174-175
> txt_out = self.cross_att(txt_output, image_output, image_output)   # 32,768
> # Lines 180-181 (CLIP-View Fusion)
> cross_feature_text = self.cross_att(text_feature, image_feature, image_feature)  # 32,768
> Here, `txt_output` and `text_feature` follow these pipelines:
> 
> ```
> # Line 132
> # 'text_features' ARE NEVER USED. Pooling feature 'text_feature' are used instead.
> text_features = output['text_model_output']['last_hidden_state']  # 128，77，512
> # Line 134
> text_feature = output['text_model_output']['pooler_output'] # 64，512
> 
> # Line 153-154
> # 'last_hidden_states' ARE NEVER USED. Pooling feature 'pooler_outputs' are used instead.
> last_hidden_states = outputs_bert.last_hidden_state  # 64,56,768
> pooler_outputs = outputs_bert.pooler_output  # 64,768
> 
> # Line 165-166
> # `squeeze` explicitly discards the sequence dimension.
> txt_output = self.text_cross(tgt=text_feature2, memory=txt_cat)  # 32,1,768
> txt_output = txt_output.squeeze(1)
> ```
> 
> Because the inputs are pooled global vectors (`[B, D]`) rather than sequences (`[B, L, D]`), the attention mechanism calculates the similarity between Sample i 's text and Sample j 's image (where j ranges over the entire batch).
> 
> * **Conceptual Implication:** The feature representation for Sample 
>     A
>    is a weighted sum of the images of Sample 
>     A
>     ,
>     B
>     ,
>     C
>     ,
>     .
>     .
>     .
>    in the current batch.
> * **Inference Violation:** In a real-world scenario (or Batch Size = 1), a tweet's sarcasm should be determined solely by its own text and image. However, this model requires a "batch context" to function as trained.
> 
> This implementation directly contradicts the methodology described in the paper.
> 
> * **Paper Claim (Section 4.3):** _"We use two cross-attention modules to compute the interactions **between the modalities**... allows for a more comprehensive consideration of multimodal information."_
> * **Reality:** The interaction is **Inter-sample** (Text of Sample 
>     i
>    
>     ↔
>    Images of Batch 
>     0.
>     .
>     .
>     N
>   ). The paper does not mention using batch-level statistics or contrastive mechanisms that would justify this behavior.
> 
> ## Experiment
> If the model learns true features, Test BS should not affect accuracy. If it relies on batch interaction, performance will collapse when Test BS differs from Train BS (especially at BS=1).
> 
> Results (All Experiments are conducted under original experiment setting in `main.py` and `train.sh`):
> 
> Training Setting	Testing Setting	Acc (%)	F1 (%)	Observation
> BS = 32	BS = 32	88.54	88.44	Reported setting in the paper and `train.sh`. Lower than results reported in the paper, but still much higher than baseline models.
> BS = 32	BS = 2	85.22	85.16	Significant degradation
> **BS = 32**	**BS = 1**	**80.24**	**80.22**	Significant degradation.
> BS = 128	BS = 128	82.36	82.34	Significant degradation
> BS = 128	BS = 32	81.53	81.43	Significant degradation
> When Batch Size = 1, `softmax` over the batch dimension results in a weight of 1.0 for the single sample, completely removing the "interaction" capability the model learned to rely on during training.
> 
> ## Conclusion
> **Avoid Using this Code as a Baseline**. Since the model violates the basic I.I.D. (Independent and Identically Distributed) assumption for inference, building upon this architecture may lead to theoretically unsound research. Any performance metrics reported with a Batch Size > 1 should not be considered representative of the model's actual predictive power. Attempting to reproduce or build upon these results is likely to yield unstable and misleading outcomes, as the model's reported superiority relies on a methodological flaw that cannot be generalized to standard application scenarios.

Appreciation for your detailed analysis and feedback

We would like to express our sincere gratitude for your keen observation and the time you took to examine our code rigorously. Feedback like this is essential for improving the quality of research within the community.

After carefully examining the logic in model.py and the CrossAttention module you pointed out, we confirm that your assessment is correct. Since the inputs are pooled features rather than sequence features, the implementation in the current uploaded code version indeed performs interaction along the batch dimension. We realize that there is a discrepancy between this implementation and the intra-sample attention mechanism described in our paper.

We attach great importance to your feedback. We are currently conducting a thorough review of the codebase to organize the correct versions. We plan to update the repository with the necessary revisions and instructions soon.

Thank you again for pointing this out. It is very helpful!

Best regards.

**Comment by Issac304** [comment]
Indeed, when reading the paper, I also noticed that the cross-sample interaction was not reasonable.

**Comment by Issac304** [comment]
@SongshuL3 Thank you for pointing out the problem in such detail.
